### DISCUSSION TITLE
Tasks taking extra 20 minutes after its real work is done

### DISCUSSION THREAD

**Discussion by Xavi-Messi** [apache/airflow#35300] [Q&A]
Hi, I'm using some BashOperator and PythonOperator tasks in my airflow dag. The dag is relatively big, it has a few thousand tasks. Somehow I see after the real work is done in the task, the tasks sometimes may take a few seconds to more than 20 minutes to finish these two last steps before finally exiting the task: 
(1) {local_task_job.py:212} INFO - Task exited with return code 0
(2) {taskinstance.py:2596} INFO - 0 downstream tasks scheduled from follow-on schedule check

Can any one tell why sometimes it needs so long for these two steps to finish? It happens more often when I have 100-200 tasks in the same task_group starting together. These tasks are running for different small datasets and they're basically doing similar work. Though each task may take less than 1 minute to finish the underlying real work, the total time consumption can be much longer due to the time consumed in the above two steps. The extra time used in these two steps for each task is different, some tasks take 1 or 2 seconds, some tasks may take 20 minutes or more.

I checked the airflow source code a bit, in step (2), it seems to lock the dag database and tries to do some 'mini-scheduling' work, is the extra time spent here related to locking the dag database? Also is there anything I can do to reduce the extra time here, it really makes the whole running process much longer than expected and is very frustrating.

Thank you very much!

**Comment by Taragolis** [comment]
Let’s start from the beginning, with try get more details about your installation and usage of Airflow

it would be nice if you provide additional info
- Version of Airflow
- DB Backend Type
- Executor Type
- Is it self hosted or managed?
- Deployment details
- how you create a DAG, is any kind of dynamic DAG generation is used? 
- Do you use any callbacks?
- Any attempt to call to airflow db into the task or callbacks? E.g. change states or something similar

**Comment by Xavi-Messi** [comment]
@Taragolis thanks for your comments, here're some details:

- **Version of Airflow**: 2.5.3
- **DB Backend Type**: PostgresSQL
- **Executor Type**: Celery
- **Is it self hosted or managed**: self hosted
- **Deployment details**: we have two scheduler running on different machines, and have 4 worker nodes. They're all running in containers.
- **how you create a DAG, is any kind of dynamic DAG generation is used**: it's using a dag file for each dag, not using dynamic dag generation.
- **Do you use any callbacks**: if you mean the callback function after the task is done, like on_success_callback, then we don't use it.
- **Any attempt to call to airflow db into the task or callbacks? E.g. change states or something similar**: This is a little complicated, say there're 3 types of task groups that all have the delay, we may call them A, B, C. The order is like A > B > C. Here some tasks in B will change the states of some other upstream tasks also in B, but A and C won't call to airflow db or change states. Type A tasks are just a BashOperator running a command line. C task is using a PythonOperator to analyze our data. I can see the extra delay when A tasks are running, while B and C tasks are in no_status. When B or C are running, while upstream tasks all done, there can also be delays.

Can you please let me know what are the possible reasons that can cause the delay? Do you suggest anything we can check on our side? Also will any interactions between different dags and dag runs cause such a delay? I see the row_locks of airflow dag run in step (2), will it cause a problem if many tasks are running at the same time? 

Also is there a best practice if we start more than 100 parallel tasks in one task group together? These tasks are very similar and just work on different datasets in parallel. Shall we just throw them out and start running all at the same time? Or is there a way to let the scheduler know that all these tasks can be run in the same worker right away after one finishes, so that they can be scheduled in the mini-scheduler in a more efficient way. It seems some of these tasks are already scheduled by the mini-scheduler
